### ISSUE TITLE
Submission: <Group 5: Bank Marketing Analysis>

### ISSUE THREAD

**Issue by nassimgha** [UBC-MDS/data-analysis-review-2023#26]
**Submitting authors:**  <[zhang-shizhe](https://github.com/zhang-shizhe)> <[Marcony1](https://github.com/Marcony1)> <[celestezhao](https://github.com/celestezhao)> <[nassimgha](https://github.com/nassimgha)>

**[Repository](https://github.com/UBC-MDS/bank-marketing-analysis):**
**[Report link](https://github.com/UBC-MDS/bank-marketing-analysis/blob/main/report/_build/html/bank_marketing_analysis_report.html):**
**Abstract/executive summary:**
In this analysis, we attempt to build a predictive model aimed at determining whether a client will subscribe to a term deposit, utilizing the data associated with direct marketing campaigns, specifically phone calls, in a Portuguese banking institution.

After exploring on several models (logistic regression, KNN, decision tree, naive Bayers), we have selected the logistic regression model as our primary predictive tool. The final model performs fairly well when tested on an unseen dataset, achieving the highest AUC (Area Under the Curve) of 0.899. This exceptional AUC score underscores the model's capacity to effectively differentiate between positive and negative outcomes. Notably, certain factors such as last contact duration, last contact month of the year and the clients' types of jobs play a significant role in influencing the classification decision.

The dataset used in this project originates from the Bank Marketing dataset created by S. Moro, P. Rita and P. Cortez at Iscte - University Institute of Lisbon. This dataset is accessible through the UCI Machine Learning Repository and can be accessed [here](https://archive.ics.uci.edu/dataset/222/bank+marketing). Among the four available datasets, we have utilized [bank-full.csv](https://archive.ics.uci.edu/static/public/222/data.csv) which contains all examples and 17 inputs. Each row in the dataset represents an individual client data including the personal details (e.g., age, occupation, loan status, etc.), information regarding their response to the marketing campaign (e.g., outcomes of the previous marketing campaign, number of contacts made during the current campaign, etc.), and the eventual subscription outcome for the term deposit.

**Editor:** @ttimbers 
**Reviewer:** Beth Ou Yang, Michelle Hunn, Gretel Tan, Julia Everitt

- [ ] I agree to abide by [MDS's Code of Conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by beth-ouyang** [comment]
## Data analysis review checklist

### Reviewer: @beth-ouyang

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [x] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [x] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 1 hr

### Review Comments: 

- The project is easily to reproduce, and the analysis is clear, well-crafted and insightful! The EDA part is really thorough, and the discussion parts after trying multiple methods are really impressive.
- Perhaps including a brief conclusion in the EDA part may be helpful for readers to understand which variables are significant.

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by gtmx23** [comment]
## Data analysis review checklist

### Reviewer: @gtmx23 

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] Repository: Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] License: Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] Installation instructions: Is there a clearly stated list of dependencies? 
- [x] Example usage: Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] Functionality documentation: Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] Readability: Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] Style guidelides: Does the code adhere to well known language style guides?
- [x] Modularity: Is the code suitably abstracted into scripts and functions?
- [x] Tests: Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] Data: Is the raw data archived somewhere? Is it accessible?
- [x] Computational methods: Is all the source code required for the data analysis available?
- [x] Conditions: Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [ ] Automation: Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] Authors: Does the report include a list of authors with their affiliations?
- [x] What is the question: Do the authors clearly state the research question being asked?
- [x] Importance: Do the authors clearly state the importance for this research question?
- [x] Background: Do the authors provide sufficient background information so that readers can understand the report?
- [x] Methods: Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] Results: Do the authors clearly communicate their findings through writing, tables and figures?
- [x] Conclusions: Are the conclusions presented by the authors correct? 
- [x] References: Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] Writing quality: Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 1.5 h

### Review Comments: 
- Scripts were very well documented - the docstrings were detailed and helpful.
- I like the use of many models - it shows that the analysis was rigorous and that the team put in a lot of effort into analysis.
- Overall, I thought the report was well written and the analysis were well done! My main complaints were with regard to the running of the scripts, but these were all minor complaints which can be easily fixed. Good job :)

**Minor Issues with Scripts**
- README.md document is well written! However, there was an extra whitespace in the section oin the split_and_process.py documentation which I had to remove to run the script. This could be removed in future README.md
```
# Split data (train-test), process data and save preprocessorsplit
python scripts/split_and_process.py \
   --raw_data='data/raw/bank-full.csv' \
   --save_to='data/processed' \
   --preprocessor_to='results/models' \ 
   --seed=522
```
- The script fit_bank_classifier.py takes a very long time to run and produces this error:
```
/opt/conda/lib/python3.11/site-packages/sklearn/neighbors/_classification.py:233: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
```
- I was not able to finish running this script as a result. Perhaps the size of training data can be reduced to speed up the fit time.
- The script `feat_imp.py` had inconsistent input names - the README.md and documentation still use `resampled_training_data` as the input, while the actual script uses the name `transformed_training_data`, which caused some error in reproducing the analysis. This should be corrected.
- The command `docker compose exec jupyter-lab /bin/bash` did not run for me, with the error `bash: docker: command not found`. However, the command `jupyter-book build report` still worked. I would suggest removing the docker compose command in the light of this.
- The report linked in the issues leads me to the raw HTML file, not the GitHub Page. This should be corrected. Perhaps the README.md instructions should also be updated with clearer instructions on opening the HTML version of the report, to tell readers who are not so tech savvy how to open the HTML file in the container.

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by mishelly-h** [comment]
## Data analysis review checklist

### Reviewer: @mishelly-h

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [ ] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [x] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [x] **Conclusions:** Are the conclusions presented by the authors correct? 
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 2h

### Review Comments: 

Overall, I think you did a great job with the report and I enjoyed reading your analysis. In the following some suggestions how you could improve your report even further:
- Running your analysis requires entering multiple code chunks into the terminal. The reproducibility could have been improved by creating a bash script (.sh) that calls your code to run the analysis. 
- I found your tests and they look good. However, I would not find your test scenarios where you show that the functions pass the tests. 
- In the report, you have used bullet points for the discussions of each of the models. I think paragraph style could better guide of the reader through the results of each model.

In summary, you have shown that you have conducted thorough research and discovered interesting findings. 

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by juliaeveritt** [comment]
## Data analysis review checklist

### Reviewer: @juliaeveritt

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [x] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 1

### Review Comments: 

Please provide more detailed feedback here on what was done particularly well, and what could be improved. It is especially important to elaborate on items that you were not able to check off in the list above.

Great work! Your report is nicely formatted and professional (I like the color scheme you chose for the plots :D ), and your analysis methods look great. I was able to reproduce your analysis with minimal errors.

Small issues I bumped into/suggestions for improvement:
* I had an error with '--seed=522' argument in the split_process.py script. Also got the error `Error: No such option: --resampled_training_data Did you mean --transformed_training_data? bash: --seed=522: command not found` on feat_imp.py
* I think your readme.Md is very clear, but also very similar to Tiff's. It would be good to either give credit to hers or make it your own a bit more! 
* Typo with 'subscription' on your report title
* Putting your discussion into paragraphs instead of bullet points would be good to maintain the professional tone of the rest of the report

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.
